### ISSUE TITLE
Fix invalid command line parameter with --fix destroys all styles

### ISSUE THREAD

**Issue by ZhangYiJiang** [stylelint/stylelint#2991]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Bug, probably caused by incorrect interaction between the arg parser and autofix

> Which rule, if any, is this issue related to?

None

> What CSS is needed to reproduce this issue?

I have a repo here: https://github.com/ZhangYiJiang/stylelint-autofix-cli-bug

> What stylelint configuration is needed to reproduce this issue?

`stylelint-config-standard` is used, although I do not think it is relevant. 

> Which version of stylelint are you using?

8.2.0 (Latest from npm)

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI - `stylelint src/*.css --fix src/` (notice the extraneous incorrect parameter `src/`) 

For a more realistic setup (which is how I discovered this bug), a user would add `stylelint src/*.css` to their `package.json`'s script section as a shortcut which they can run using `yarn lint` (or `npm run lint`, although this bug doesn't seem to repro with npm, maybe due to how npm handles parameters). If they then try to run `--fix` and accidentally add an incorrect parameter, the bug is triggered. 

``` shell 
$ cat src/app.css 
#app {
  background: orange;
}

$ yarn lint --fix src/
yarn run v1.2.1
$ stylelint 'src/*.css' --fix src/
Error: EISDIR: illegal operation on a directory, read
    at Error (native)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

$ git diff 
diff --git a/src/app.css b/src/app.css
index 52950c1..e69de29 100644
--- a/src/app.css
+++ b/src/app.css
@@ -1,3 +0,0 @@
-#app {
-  background: orange;
-}
``` 

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

> What did you expect to happen?

Error about invalid command line options 

> What actually happened (e.g. what warnings or errors you are getting)?

Error about invalid command line options, *and* all linted styles are destroyed

**Comment by jeddy3** [comment]
@ZhangYiJiang Thanks for the using the template and for the detailed write-up!

It could be an interesting task to track down what's going on here. Does it only destroy the styles if the incorrect additional parameter is a valid path?

Anyhow, I'll label up as a bug and help wanted. @ZhangYiJiang Feel free to dig into the code yourself if you'd like to investigate what is causing this.

**Comment by ZhangYiJiang** [comment]
It seems that having an incorrect path causes Stylelint to exit correctly 

```
$ yarn lint --fix /nope
yarn run v1.2.1
$ stylelint 'src/*.css' --fix /nope
Done in 1.29s.
``` 

I suspect the exception caused Stylelint to somehow not write the linted results back, but it's quite odd.

**Comment by jeddy3** [comment]
>  but it's quite odd.

Yes, it's very odd.  

> It seems that having an incorrect path causes Stylelint to exit correctly

Good to know. I think a logical first place to start is to follow the input glob through the system (from the CLI down to the writing the transformed styles back).

**Comment by ZhangYiJiang** [comment]
I have a hypothesis - this is caused by a combination of two bugs: 

1. Entering in a directory crashes Stylelint - not found out why yet, but it's quite obvious from the output  
2. Stylelint uses [`fs.writeFile`](https://github.com/stylelint/stylelint/blob/master/lib/standalone.js#L214) to write the fixed file back, which is not atomic - if the program crashes or exits while while this operation is running, the file is left in a truncated state. This is *really* bad here because Node lets you start hundreds of these operations, and since `writeFile` is a truncating write, a crash immediately after this will truncate *all* of the linted sources. 

I'll work on fixing (2) - the correct way to handle this is to write to a temporary file first, then replace the original with the temporary file (https://stackoverflow.com/questions/1812115/how-to-safely-write-to-a-file). There's also some room for optimization here - Stylelint shouldn't even need to touch the files if there were no changes.

**Comment by jeddy3** [comment]
> I have a hypothesis - this is caused by a combination of two bugs:

Nice sleuthing.

> I'll work on fixing (2)

That would be awesome!

**Comment by ZhangYiJiang** [comment]
@jeddy3 I have a fix for (2) in #2992, but it triggers a memory leak in `jest`/`graceful-fs`. I've looked at a few ways to work around this, but the best way (keep a fork of [`write-file-atomic`](https://www.npmjs.com/package/write-file-atomic) with `graceful-fs` swapped for `fs`) is still pretty bad. Can you see if there's a better way to do this?

**Comment by jeddy3** [comment]
> I have a fix for (2) in #2992,

Fantastic.

> Can you see if there's a better way to do this?

Umm, nothing comes to mind. This is quite a nasty bug and I think it's worth maintaining a fork if that resolves the issue. @stylelint/core Any objections to maintaining a fork in the stylelint org? 

> There's also some room for optimization here - Stylelint shouldn't even need to touch the files if there were no changes.

It looks like this has been raised as a separate issue: https://github.com/stylelint/stylelint/issues/3003 - it was best to handle this optimisation in a separate PR anyway.

(Sorry about the delay getting back to you here and in your PR - it was a busy week).

**Comment by CAYdenberg** [comment]
> Any objections to maintaining a fork in the stylelint org?

No objections, but should we file a bug in jest/graceful-fs? Or do we not really know the cause of the memory leak?

**Comment by ZhangYiJiang** [comment]
Issues have been filed the last time this has came up - 

- graceful-fs: https://github.com/isaacs/node-graceful-fs/issues/102
- Jest: https://github.com/facebook/jest/issues/2179 
- tmp (which also has a similar memory leak when used with Jest): https://github.com/raszi/node-tmp/issues/129

**Comment by jeddy3** [comment]
> graceful-fs: isaacs/node-graceful-fs#102

As that issue was never addressed, I think forking might be the best way forward.

**Comment by ZhangYiJiang** [comment]
@jeddy3 If it's not too much trouble, the PR is ready for re-review
